File: shuffle_cards/deck.rb

#initialize a deck of cards and print out
#shuffle deck

class Deck
  
  attr_reader :suits, :cards, :length

  def initialize
    @suits = ['Clubs', 'Diamonds', 'Hearts', 'Spades']
    @cards = build
    @length = 52
  end

  def shuffle
    # iterate over each card and swap it with a randomly selected card
    # each card has the same probability of landing in any given spot (1/52)
    cards.each_with_index do |card, current_index|
      new_index = random
      cards[current_index], cards[new_index] = cards[new_index], cards[current_index]
    end
    cards
  end

  private

  def random
    rand(0...length)
  end

  def build
    new_deck = []
    suits.each do |suit|
      for card in (2..14)
        case card
        when 11
          new_deck << "Jack of #{suit}"
        when 12
          new_deck << "Queen of #{suit}"
        when 13
          new_deck << "King of #{suit}"
        when 14
          new_deck << "Ace of #{suit}"
        else 2..10
          new_deck << "#{card} of #{suit}"
        end
      end
    end
    new_deck
  end
end

deck = Deck.new
puts deck.cards
puts deck.length
puts deck.shuffle
puts deck.length

File: shuffle_cards_js/src/Deck.js
const suits = ["hearts", "diamonds", "clubs", "spades"];

class Deck {

	newDeck() {
		// iterate over suits
		// iterate over cards
		// for each suit push card in Deck
		// return deck
		let deck = [];
		for (var i = 0; i < suits.length; i++) {
			for (var j = 2; j < 15; j++) {
				if (j < 11) {
					deck.push(j + "of " + suits[i]);
				} else if (j == 11) {
					deck.push("Jack of " + suits[i]);
				} else if (j == 12) {
					deck.push("Queen of " + suits[i]);
				} else if (j == 13) {
					deck.push("Kind of " + suits[i]);
				} else { 
					deck.push("Ace of " + suits[i]);
				}
			}
		}
		return deck;
	}

	shuffle() {
		// iterate over deck
		// swap cards with a randomly generated val btwn 0-51
		let deck = this.newDeck();
		for (var i = 0; i < deck.length; i++) {
			let random = this.getRandomIntInclusive(1, 51);
			deck[i], deck[random] = deck[random], deck[i];
		}
		return deck;
	}

	// Returns a random integer between min (included) and max (included)
	// Using Math.round() will give you a non-uniform distribution!
	getRandomIntInclusive(min, max) {
		return Math.floor(Math.random() * (max - min + 1) + min);
	}

}

Deck;
